File: ios/KeyWitness/Keyboard/DIDKey.swift
import Foundation

/// did:key encoding for Ed25519 public keys.
///
/// Format: did:key:z<base58btc(0xed01 || raw_public_key)>
///
/// Uses base58btc encoding (Bitcoin alphabet).
enum DIDKey {

    // MARK: - Public API

    /// Encode an Ed25519 public key (32 bytes) as a did:key identifier.
    static func ed25519ToDIDKey(_ publicKey: Data) -> String {
        precondition(publicKey.count == 32, "Ed25519 public key must be 32 bytes")
        var prefixed = Data([0xed, 0x01])
        prefixed.append(publicKey)
        return "did:key:z\(Base58.encode(prefixed))"
    }

    /// Build the verification method ID: did:key:z...#z...
    static func verificationMethodId(for did: String) -> String {
        let fragment = String(did.dropFirst("did:key:".count))
        return "\(did)#\(fragment)"
    }
}

// MARK: - Base58 (Bitcoin alphabet)

/// Minimal Base58 implementation using the Bitcoin alphabet.
/// Used for multibase 'z' prefix encoding in did:key identifiers.
enum Base58 {

    private static let alphabet = Array("123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz")

    /// Encode raw bytes to base58btc string.
    static func encode(_ data: Data) -> String {
        var bytes = [UInt8](data)

        // Count leading zeros
        var leadingZeros = 0
        for byte in bytes {
            if byte == 0 { leadingZeros += 1 }
            else { break }
        }

        // Convert to base58
        var result: [Character] = []
        while !bytes.isEmpty {
            var carry = 0
            var newBytes: [UInt8] = []
            for byte in bytes {
                carry = carry * 256 + Int(byte)
                if !newBytes.isEmpty || carry / 58 > 0 {
                    newBytes.append(UInt8(carry / 58))
                }
                carry = carry % 58
            }
            result.append(alphabet[carry])
            bytes = newBytes
        }

        // Add leading '1's for each leading zero byte
        for _ in 0..<leadingZeros {
            result.append(alphabet[0])
        }

        return String(result.reversed())
    }

    /// Decode a base58btc string to raw bytes.
    static func decode(_ string: String) -> Data? {
        let chars = Array(string)
        var result: [UInt8] = []

        // Count leading '1's (maps to 0x00 bytes)
        var leadingOnes = 0
        for char in chars {
            if char == alphabet[0] { leadingOnes += 1 }
            else { break }
        }

        for char in chars {
            guard let index = alphabet.firstIndex(of: char) else {
                return nil // Invalid character
            }
            var carry = index
            for i in stride(from: result.count - 1, through: 0, by: -1) {
                carry += Int(result[i]) * 58
                result[i] = UInt8(carry & 0xFF)
                carry >>= 8
            }
            while carry > 0 {
                result.insert(UInt8(carry & 0xFF), at: 0)
                carry >>= 8
            }
        }

        // Prepend zero bytes for leading '1's
        let zeros = [UInt8](repeating: 0, count: leadingOnes)
        return Data(zeros + result)
    }
}

File: web/src/lib/didkey.ts
/**
 * did:key encoding/decoding for Ed25519 and P-256 public keys.
 *
 * did:key is a self-resolving DID method — no registry or blockchain needed.
 * Format: did:key:<multibase(multicodec-prefix || raw-public-key)>
 *
 * Multicodec prefixes (varint-encoded):
 *   Ed25519: 0xed 0x01
 *   P-256:   0x80 0x24
 *
 * Multibase prefix: z = base58btc
 *
 * @module
 */

import bs58 from "bs58";

// Multicodec prefixes (varint-encoded)
const ED25519_PREFIX = new Uint8Array([0xed, 0x01]);
const P256_PREFIX = new Uint8Array([0x80, 0x24]);

/**
 * Encode an Ed25519 public key (32 bytes) as a did:key identifier.
 *
 * @example
 * ed25519ToDIDKey(publicKeyBytes) // "did:key:z6MkhaXg..."
 */
export function ed25519ToDIDKey(publicKey: Uint8Array): string {
  if (publicKey.length !== 32) {
    throw new Error(`Ed25519 public key must be 32 bytes, got ${publicKey.length}`);
  }
  const prefixed = new Uint8Array(ED25519_PREFIX.length + publicKey.length);
  prefixed.set(ED25519_PREFIX, 0);
  prefixed.set(publicKey, ED25519_PREFIX.length);
  return `did:key:z${bs58.encode(prefixed)}`;
}

/**
 * Encode a P-256 public key (33 bytes compressed or 65 bytes uncompressed) as did:key.
 */
export function p256ToDIDKey(publicKey: Uint8Array): string {
  if (publicKey.length !== 33 && publicKey.length !== 65) {
    throw new Error(`P-256 public key must be 33 or 65 bytes, got ${publicKey.length}`);
  }
  const prefixed = new Uint8Array(P256_PREFIX.length + publicKey.length);
  prefixed.set(P256_PREFIX, 0);
  prefixed.set(publicKey, P256_PREFIX.length);
  return `did:key:z${bs58.encode(prefixed)}`;
}

/**
 * Decode a did:key identifier to extract the raw public key bytes and key type.
 */
export function decodeDIDKey(did: string): { keyType: "Ed25519" | "P-256"; publicKey: Uint8Array } {
  if (!did.startsWith("did:key:z")) {
    throw new Error(`Invalid did:key format: ${did}`);
  }
  const multibaseBody = did.slice("did:key:z".length);
  const decoded = bs58.decode(multibaseBody);

  if (decoded[0] === 0xed && decoded[1] === 0x01) {
    return { keyType: "Ed25519", publicKey: decoded.slice(2) };
  }
  if (decoded[0] === 0x80 && decoded[1] === 0x24) {
    return { keyType: "P-256", publicKey: decoded.slice(2) };
  }
  throw new Error(`Unknown multicodec prefix: 0x${decoded[0].toString(16)} 0x${decoded[1].toString(16)}`);
}

/**
 * Build the verification method ID from a did:key.
 * Format: did:key:z6Mk...#z6Mk...
 */
export function verificationMethodId(did: string): string {
  const fragment = did.slice("did:key:".length);
  return `${did}#${fragment}`;
}
